emperor/core.py

```diff
@@ -302,39 +303,43 @@ def _repr_html_(self):
 
         return display(HTML(str(self)))
 
-    def _validate_metadata(self, metadata, matrix, ignore_missing_samples):
+    def _validate_metadata(self, metadata, matrix, ignore_missing_samples,
+                           kind='sample'):
+
+        if kind not in {'sample', 'feature'}:
+            raise ValueError('Unsupported "kind" value %s' % kind)
 
         # metadata is optional for biplots, so we just create an empty table
         if metadata is None:
             metadata = pd.DataFrame(index=pd.Index(matrix.index, name='id'))
-            metadata['all'] = 'All objects'
+            metadata['all'] = 'All elements'
             return metadata
 
-        ordination_samples = set(matrix.index)
-        difference = ordination_samples - set(metadata.index)
+        ordination_elements = set(matrix.index)
+        difference = ordination_elements - set(metadata.index)
 
-        if difference == ordination_samples:
-            raise ValueError('None of the sample identifiers match between the'
+        if difference == ordination_elements:
+            raise ValueError('None of the %s identifiers match between the'
                              ' metadata and the coordinates. Verify that you '
                              'are using metadata and coordinates corresponding'
-                             ' to the same dataset.')
+                             ' to the same dataset.' % kind)
 
         if difference and not ignore_missing_samples:
-            raise KeyError("There are samples not included in the mapping "
+            elements = ', '.join(sorted([str(i) for i in difference]))
+            raise KeyError("There are %ss not included in the %s mapping "
                            "file. Override this error by using the "
                            "`ignore_missing_samples` argument. Offending "
-                           "samples: %s"
-                           % ', '.join(sorted([str(i) for i in difference])))
+                           "%ss: %s" % (kind, kind, kind, elements))
         elif difference and ignore_missing_samples:
-            warnings.warn("%d out of %d samples have no metadata and are being"
+            warnings.warn("%d out of %d %ss have no metadata and are being"
                           " included with a placeholder value." %
-                          (len(difference), len(ordination_samples)),
+                          (len(difference), len(ordination_elements), kind),
                           EmperorWarning)
 
-            # pad the missing samples
+            # pad the missing elements
             data = np.full((len(difference), metadata.shape[1]),
-                           'This sample has no metadata', dtype='<U27')
-            pad = pd.DataFrame(index=difference, columns=self.mf.columns,
+                           'This element has no metadata', dtype='<U28')
```

Comment:
I *think* just using a dtype of str will allow numpy to determine the exact type it wants to use?

Reply:
Good point, come to think about it I can do a fillna using the intermediate DataFrame, fixed - thanks!